pkg/executor/join/hash_join_base.go

```diff
@@ -0,0 +1,359 @@
+	JoinType      plannercore.JoinType
+	IsNullAware   bool
+	memTracker    *memory.Tracker // track memory usage.
+	diskTracker   *disk.Tracker   // track disk usage.
+}
+
+type probeSideTupleFetcherBase struct {
+	ProbeSideExec      exec.Executor
+	probeChkResourceCh chan *probeChkResource
+	probeResultChs     []chan *chunk.Chunk
+	requiredRows       int64
+	joinResultChannel  chan *hashjoinWorkerResult
+}
+
+func (fetcher *probeSideTupleFetcherBase) initializeForProbeBase(concurrency uint, joinResultChannel chan *hashjoinWorkerResult) {
+	// fetcher.probeResultChs is for transmitting the chunks which store the data of
+	// ProbeSideExec, it'll be written by probe side worker goroutine, and read by join
+	// workers.
+	fetcher.probeResultChs = make([]chan *chunk.Chunk, concurrency)
+	for i := uint(0); i < concurrency; i++ {
+		fetcher.probeResultChs[i] = make(chan *chunk.Chunk, 1)
+	}
+	// fetcher.probeChkResourceCh is for transmitting the used ProbeSideExec chunks from
+	// join workers to ProbeSideExec worker.
+	fetcher.probeChkResourceCh = make(chan *probeChkResource, concurrency)
+	for i := uint(0); i < concurrency; i++ {
+		fetcher.probeChkResourceCh <- &probeChkResource{
+			chk:  exec.NewFirstChunk(fetcher.ProbeSideExec),
+			dest: fetcher.probeResultChs[i],
+		}
+	}
+	fetcher.joinResultChannel = joinResultChannel
+}
+
+func (fetcher *probeSideTupleFetcherBase) handleProbeSideFetcherPanic(r any) {
+	for i := range fetcher.probeResultChs {
+		close(fetcher.probeResultChs[i])
+	}
+	if r != nil {
+		fetcher.joinResultChannel <- &hashjoinWorkerResult{err: util.GetRecoverError(r)}
+	}
+}
+
+type isBuildSideEmpty func() bool
+
+func wait4BuildSide(isBuildEmpty isBuildSideEmpty, canSkipIfBuildEmpty bool, hashJoinCtx *hashJoinCtxBase) (skipProbe bool, err error) {
+	select {
+	case <-hashJoinCtx.closeCh:
+		return true, nil
+	case err := <-hashJoinCtx.buildFinished:
+		if err != nil {
+			return false, err
+		}
+	}
+	if isBuildEmpty() && canSkipIfBuildEmpty {
+		return true, nil
+	}
+	return false, nil
+}
+
+// fetchProbeSideChunks get chunks from fetches chunks from the big table in a background goroutine
+// and sends the chunks to multiple channels which will be read by multiple join workers.
+func (fetcher *probeSideTupleFetcherBase) fetchProbeSideChunks(ctx context.Context, maxChunkSize int, isBuildEmpty isBuildSideEmpty, canSkipIfBuildEmpty, needScanAfterProbeDone, shouldLimitProbeFetchSize bool, hashJoinCtx *hashJoinCtxBase) {
+	hasWaitedForBuild := false
+	for {
+		if hashJoinCtx.finished.Load() {
+			return
+		}
+
+		var probeSideResource *probeChkResource
+		var ok bool
+		select {
+		case <-hashJoinCtx.closeCh:
+			return
+		case probeSideResource, ok = <-fetcher.probeChkResourceCh:
+			if !ok {
+				return
+			}
+		}
+		probeSideResult := probeSideResource.chk
+		if shouldLimitProbeFetchSize {
+			required := int(atomic.LoadInt64(&fetcher.requiredRows))
+			probeSideResult.SetRequiredRows(required, maxChunkSize)
+		}
```

Comment:
```suggestion
func (fetcher *probeSideTupleFetcherBase) getProbeChkResource(hashJoinCtx *hashJoinCtxBase, shouldLimitProbeFetchSize bool, maxChunkSize int) *probeChkResource {
	if hashJoinCtx.finished.Load() {
		return nil
	}

	var probeSideResource *probeChkResource
	var ok bool
	select {
	case <-hashJoinCtx.closeCh:
		return nil
	case probeSideResource, ok = <-fetcher.probeChkResourceCh:
		if !ok {
			return nil
		}
	}

	if shouldLimitProbeFetchSize {
		required := int(atomic.LoadInt64(&fetcher.requiredRows))
		probeSideResource.chk.SetRequiredRows(required, maxChunkSize)
	}
	return probeSideResource
}

probeSideResource := fetcher.getgetProbeChkResource(hashJoinCtx, shouldLimitProbeFetchSize, maxChunkSize)
if probeSideResource == nil {
	return
}
```
Maybe we can encapsulate these codes into a single function named like `getgetProbeChkResource` and directly return when `getgetProbeChkResource` returns nil.

Reply:
Done